pkg/kvstore/allocator/allocator_test.go

```diff
@@ -111,6 +115,117 @@ func (s *AllocatorSuite) BenchmarkAllocate(c *C) {
 
 }
 
+func (s *AllocatorSuite) TestRunLocksGC(c *C) {
+	allocatorName := randomTestName()
+	maxID := idpool.ID(256 + c.N)
+	// FIXME: Did this previousy use allocatorName := randomTestName() ? so TestAllocatorKey(randomeTestName())
+	backend1, err := NewKVStoreBackend(allocatorName, "a", TestAllocatorKey(""))
+	c.Assert(err, IsNil)
+	c.Assert(err, IsNil)
+	allocator, err := allocator.NewAllocator(TestAllocatorKey(""), backend1, allocator.WithMax(maxID), allocator.WithoutGC())
+	c.Assert(err, IsNil)
+	shortKey := TestAllocatorKey("1;")
+	var (
+		lock1, lock2 kvstore.KVLocker
+		gotLock1     = make(chan struct{})
+		gotLock2     = make(chan struct{})
+	)
+	go func() {
+		var (
+			err error
+		)
+		lock1, err = backend1.Lock(context.Background(), shortKey)
+		c.Assert(err, IsNil)
+		close(gotLock1)
+		var client kvstore.BackendOperations
+		switch s.backend {
+		case "etcd":
+			client, _ = kvstore.NewClient(context.Background(),
+				s.backend,
+				map[string]string{
+					kvstore.EtcdAddrOption: kvstore.EtcdDummyAddress(),
+				},
+				nil,
+			)
+		case "consul":
+			client, _ = kvstore.NewClient(context.Background(),
+				s.backend,
+				map[string]string{
+					kvstore.ConsulAddrOption:   kvstore.ConsulDummyAddress(),
+					kvstore.ConsulOptionConfig: kvstore.ConsulDummyConfigFile(),
+				},
+				nil,
+			)
+		}
+		lock2, err = client.LockPath(context.Background(), allocatorName+"/locks/"+shortKey.GetKey())
+		c.Assert(err, IsNil)
+		close(gotLock2)
+	}()
+	staleLocks := map[string]kvstore.Value{}
+	staleLocks, err = allocator.RunLocksGC(staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 0)
+
+	// Wait until lock1 is gotten.
+	c.Assert(testutils.WaitUntil(func() bool {
+		select {
+		case <-gotLock1:
+			return true
+		default:
+			return false
+		}
+	}, 5*time.Second), IsNil)
+
+	// wait until lock2 is hold
+	time.Sleep(time.Second)
+
+	// Check which locks are stale, it should be lock1 and lock2
+	staleLocks, err = allocator.RunLocksGC(staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 2)
+
+	// Get the oldest lock being held
+	var (
+		oldestRev     uint64
+		oldestLeaseID int64
+		sessionID     string
+	)
+	for _, v := range staleLocks {
+		if v.ModRevision < oldestRev {
+			oldestRev = v.ModRevision
+			oldestLeaseID = v.LeaseID
+			sessionID = v.SessionID
+		}
+	}
+	staleLocks[allocatorName+"/locks/"+shortKey.GetKey()] = kvstore.Value{
+		ModRevision: oldestRev,
+		LeaseID:     oldestLeaseID,
+		SessionID:   sessionID,
+	}
+
+	// GC lock1 but we still get the lock2
+	staleLocks, err = allocator.RunLocksGC(staleLocks)
+	c.Assert(err, IsNil)
+	c.Assert(len(staleLocks), Equals, 0)
+
+	// Wait until lock2 is gotten as it should have happen since we have
+	// GC lock1.
```

Comment:
I'm confused between this and the line 179 which also appears to be trying to wait for the lock2 to be held.

Reply:
Please read my explanation I think it will become clear to you.